### PR TITLE
Move QgsBrowserDockWidget from app to gui

### PR DESCRIPTION
**Pull request by elpaso**
As mentioned in https://github.com/qgis/QGIS/pull/4629#discussion_r119371736

## TODO
- [x] Documentation of the public methods (there wasn't any)
- [x] SIP bindings ? That were also missing

### CODE REVIEW

File: src/gui/qgsbrowserdockwidget.cpp [hunk 1]
```diff
@@ -657,6 +653,11 @@ void QgsBrowserDockWidget::setFilter()
     mProxyModel->setFilter( filter );
 }
 
+void QgsBrowserDockWidget::updateProjectHome()
+{
+  if (mModel) mModel->updateProjectHome();
```
Comment (nyalldawson) [inline src/gui/qgsbrowserdockwidget.cpp]: split to multiline ( I thought astyle was supposed to pick that up?)

File: src/gui/qgsbrowserdockwidget.h [hunk 2]
```diff
@@ -134,6 +136,12 @@ class APP_EXPORT QgsBrowserDockWidget : public QgsDockWidget, private Ui::QgsBro
     void selectionChanged( const QItemSelection &selected, const QItemSelection &deselected );
     void splitterMoved();
 
+  signals:
+    // A file needs to be opened
+    void openFile( const QString );
```
Comment (nyalldawson) [inline src/gui/qgsbrowserdockwidget.h]: const QString &

File: src/gui/qgsbrowserdockwidget.h [hunk 3]
```diff
@@ -134,6 +136,12 @@ class APP_EXPORT QgsBrowserDockWidget : public QgsDockWidget, private Ui::QgsBro
...
+    // Drop uri list needs to be handled
+    void handleDropUriList( QgsMimeDataUtils::UriList );
```
Comment (nyalldawson) [inline src/gui/qgsbrowserdockwidget.h]: const ref

File: src/gui/qgsbrowserdockwidget.h [hunk 4]
```diff
@@ -134,6 +136,12 @@ class APP_EXPORT QgsBrowserDockWidget : public QgsDockWidget, private Ui::QgsBro
     void selectionChanged( const QItemSelection &selected, const QItemSelection &deselected );
     void splitterMoved();
 
+  signals:
+    // A file needs to be opened
```
Comment (nyalldawson) [inline src/gui/qgsbrowserdockwidget.h]: Usually signals are documented using the convention: "Emitted when .... "

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Thank you ! This will be very useful to have accessible outside of app

**Review by wonder-sk — APPROVED** [review]
Hi Alessandro - looks great!

I have nothing else to add in addition to Nyall's comments and your to-do list :-)

File: src/gui/qgsbrowserdockwidget.h [hunk 5]
```diff
@@ -33,23 +33,45 @@ class QgsLayerItem;
...
   public:
+
+    /**
+      * Constructor for QgsBrowserPropertiesWidget
+      * \param parent parent widget
+      */
     explicit QgsBrowserPropertiesWidget( QWidget *parent = nullptr );
-    static QgsBrowserPropertiesWidget *createWidget( QgsDataItem *item, QWidget *parent = nullptr );
+    //! Factory method to create a new browser widget
+    static QgsBrowserPropertiesWidget *createWidget( QgsDataItem *item, QWidget *parent = nullptr ) SIP_FACTORY;
```
Comment (elpaso) [inline src/gui/qgsbrowserdockwidget.h]: @wonder-sk can you please check if the factory annotation here is appropriate?

File: src/gui/qgsbrowserdockwidget.h [hunk 6]
```diff
@@ -62,94 +84,169 @@ class QgsBrowserPropertiesWidget : public QWidget
...
 {
     Q_OBJECT
   public:
+
+    /**
+      * Constructor for QgsBrowserDockWidget
+      * \param name name of the widget
+      * \param parent parent widget
+      */
     explicit QgsBrowserDockWidget( const QString &name, QWidget *parent = nullptr );
```
Comment (wonder-sk) [inline src/gui/qgsbrowserdockwidget.h]: The parent argument should have SIP_TRANSFERTHIS annotation

File: src/gui/qgsbrowserdockwidget.h [hunk 7]
```diff
@@ -62,94 +84,169 @@ class QgsBrowserPropertiesWidget : public QWidget
...
-    // A file needs to be opened
-    void openFile( const QString );
-    // Drop uri list needs to be handled
-    void handleDropUriList( QgsMimeDataUtils::UriList );
+    //! Emitted when a file needs to be opened
+    void openFile( const QString & );
+    //! Emitted when drop uri list needs to be handled
+    void handleDropUriList( const QgsMimeDataUtils::UriList & );
 
   protected:
```
Comment (wonder-sk) [inline src/gui/qgsbrowserdockwidget.h]: I think this whole section can be made private instead (apart from showEvent() which is protected from QWidget) - those methods were not really meant for re-use in further subclasses. Turning the methods to private will remove them from python interface altogether and will make the class API simpler.

File: src/gui/qgsbrowserdockwidget.h [hunk 8]
```diff
@@ -62,94 +84,169 @@ class QgsBrowserPropertiesWidget : public QWidget
...
 
   private:
     QgsBrowserPropertiesWidget *mPropertiesWidget = nullptr;
     QString mSettingsSection;
 };
 
+/**
+ * \ingroup gui
+ * The QgsBrowserDockWidget class
+ * \since QGIS 2.10
```
Comment (wonder-sk) [inline src/gui/qgsbrowserdockwidget.h]: Should read QGIS 3.0 as the class has not been available in API in earlier versions

**Review by wonder-sk — COMMENTED** [review]
I think all the classes apart from QgsBrowserDockWidget should not get exposed in the API - they are implementation details of the dock widget itself and they were not really meant for general consumption. Having those classes not exposed in public API will give us more flexibility to change the browser dock internals without the burden of having to keep API compatibility. (for that reason I omitted reviewing whether other classes have correct SIP annotations)

**Comment by wonder-sk** [comment]
Actually it would be nice to move all those internal classes to a separate "private" header file (e.g. qgsbrowserdockwidget_p.h) and they should not have GUI_EXPORT with the class definition (so they do not appear in C++ API). See for example qgscoordinatetransform_p.h

**Pull request merged**